### PR TITLE
Traffic light classifier

### PR DESCRIPTION
**Pull request by manu291**
Note: This is to be reviewed after PR #7 and PR #8 as it is based on the corresponding branches.

This detects and classifies the traffic lights on the camera image.
To test it use:
```bash
roslaunch launch/styx.launch log_level:=debug
```
Check "Camera" and uncheck "Manual". Look at the debug info in the terminal and check that it is consistent.

Note:
Initially I thought about simply fine tuning a classifier on the camera images. But it turned out it was not good enough, I still got around 10-15% of images wrongly classified. I believe this is because it is hard to classify images that vary a lot in size with just a classifier. So finally I use a detector taken from the tensorflow object detection API. It is maintained by the Tensorflow team and part of the standard Tensorflow tools.

### CODE REVIEW

**Comment by manu291** [comment]
Note: This is to be reviewed after PR #7 and PR #8 as it is based on the corresponding branches.

**Comment by Labonneguigue** [comment]
@manu291 Could you please update this branch with the latest from master please ?
Could you please specify the expected behavior in the simulator ? 
Cheers

File: ros/src/tl_detector/tl_detector.py [hunk 1]
```diff
@@ -49,13 +60,25 @@ def __init__(self):
...
+        self.get_light = rospy.get_param("/get_light").lower()
+
         rospy.spin()
 
     def pose_cb(self, msg):
         self.pose = msg
 
     def waypoints_cb(self, waypoints):
         self.waypoints = waypoints
+        self.find_stop_line_closest_waypoints()
```
Comment (Labonneguigue) [inline ros/src/tl_detector/tl_detector.py]: If there are several callbacks for the base_waypoints, the calculations for finding the traffic lights' stop lines are duplicated.

File: ros/src/tl_detector/tl_detector.py [hunk 2]
```diff
@@ -100,8 +123,20 @@ def get_closest_waypoint(self, pose):
...
-        return 0
+
+        shortest_distance = np.inf
+        closest_wp = -1
+        pos = pose.position
+        waypoints_list = self.waypoints.waypoints
+
+        for i, wp in enumerate(waypoints_list):
+            wp_pos = wp.pose.pose.position
+            distance = (wp_pos.x - pos.x) ** 2 + (wp_pos.y - pos.y) ** 2 + (wp_pos.z - pos.z) ** 2
```
Comment (Labonneguigue) [inline ros/src/tl_detector/tl_detector.py]: I feel like it is not necessary to compute this at each frame. 
We could take the final_waypoints from the waypoint updater and based upon the first value (which will be the closest_waypoint) and its distance to the next traffic light location, we could trigger the traffic light detection or not. What do you think ?

**Review by Labonneguigue — CHANGES_REQUESTED** [review]
Can't really test in the simulator. Massive latency.
I am trying to find a solution.
(2 comments)

**Comment by manu291** [comment]
@Labonneguigue Expected behavior: The debug log should output the correct state of the traffic light ahead whenever it is within 120 meters of the car and visible from the camera.

**Review by Labonneguigue — APPROVED** [review]

**Pull request merged**